proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,165 @@
+# 3rd proposal formal spec overview
+
+This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.
+
+## Abstract Syntax
+
+### Types
+
+Exception Types
+
+```
+exntype ::= [t*] -> []
+```
+
+### Instructions
+
+```
+instr ::= ... | throw x | rethrow l
+        | try bt instr* (catch x instr*)+ end
+        | try bt instr* (catch x instr*)* catch_all instr* end
+        | try bt instr* delegate l
+        | try bt instr* unwind instr* end
+```
+
+### Modules
+
+Exceptions (definitions)
+
+```
+exn ::= export* exception exntype  | export* exception exntype import
+```
+
+Modules
+
+
+```
+mod ::= module ... exn*
+```
+
+## Validation (Typing)
+
+To verify that a `try...delegate l` instruction refers to a label introduced by a try block (I'll call this a try-label), introduce a type attribute to labels in the validation context, which type attribute is set to `try` when the label is a try-label. The original notation `label [t*]` will then be a shortcut for `label {result [t*], type ε}`.
+
+Moreover, to verify that the `rethrow l` instruction refers to a label introduced by a catch block (call this a catch-label), we allow this type attribute of labels in the validation context to be set to `catch` when the label is a catch label. This addition is reflected in the execution rules, by the administrative instruction `caught` which introduces a label around the instructions of the catching try-block's catch instructions.
```

Comment:
So the intention here is, as in the case of `delegate`, `rethrow` can target any label including labels for blocks, loops, and try-delegates, but they are just validation failures, right?

The concept of catch label is a little new. Can we consider it in the same as try label? Then we have this problem:
```wasm
try $label0
  rethrow $label0  ;; cannot be done, because it's not within catch below
catch
end
```

But `delegate` has another problem anyway if it we say it takes a try block:
```wasm
try $label0
catch
  try
    ...
  delegate $label0  ;; cannot be done, because $label0's catch is not below but above here
end
```

`delegate` can be considered to target a catch label too, but in this case it can be considered to target `try`:
```wasm
try $label1
  try $label0
    try
      ...
    delegate $label0
  delegate $label1
end
```

So my question is
1. For `rethrow`: If we introduce a concept of catch label, should we also mention that that this works in the same way as its corresponding `try` block for other branch-like constructs?
2. For `delegate`: The concept of try block may not work in something like my second example, but as we can see in my third example, it is different from catch label. Should we introduce a new concept?

I might be mistaken on something, so please let me know.

Reply:
Gentle ping :)

Reply:
Sorry I'm late replying @aheejin!  

> The concept of catch label is a little new. Can we consider it in the same as try label? Then we have this problem:
> 
> ```
> try $label0
>   rethrow $label0  ;; cannot be done, because it's not within catch below
> catch
> end
> ```

I used catch labels just as a helper to make sure that the `rethrow` is inside a catch block, not as something one can use in the text format. In your example, `rethrow` is not inside a catch block, so that example would not validate.

I'm not sure I understand the problem with `delegate`, I'll try to look at this again later today or tomorrow.

Reply:
@aheejin I see now what you mean, yes, you are right. 

In the second example, the label from the catch block would interfere with `delegate`, when it shouldn't. Also `rethrow` should not interfere with other labels, it should just count the stack of nested catch blocks it is in - to rethrow an exception, an exception must be caught, so we can only `rethrow n` the exception caught by the `n`th surrounding catch block.

I think to solve this, the catch-labels should be removed, and a separate concept of a `caught`-stack should be added to the runtime, described by a parallel to blocks:

```
C^0 ::= val^m [_] instr
C^{n+1} ::= caught{exnaddr val^k} C^n end
```
Also a new stack `caught` to validation contexts, which gets an element "`caught`" prepended whenever we catch an exception by a `catch` block.

About the 3rd example, did you omit a catch in the outer try block on purpose, meaning that `try`s without catches could be allowed, as @tlively suggested in #131 ? In that case, in your 3rd example, if we replace `...` by `throw $exn` then the example would reduce to `throw $exn`. I think it would be a nice simplification to the syntax to allow catchless try blocks. They are equivalent to `try ... catch_all rethrow 0 end` AFAICT, so no new behaviour would be introduced with this simplification.

If we additionally allow `delegate` to target any label, as @RossTate [suggested](https://github.com/WebAssembly/exception-handling/pull/143/files#r522996966), then we can have the same effect also without a try block, as he noted, and we wouldn't have to make any changes to labels. I admit I am not so fond of adding this auxiliary information to labels, I just thought I had to. 

To clarify I'll commit the changes I'm suggesting later today, in three separate patches:

1. add `caught`-stack to the runtime and to validation contexts and remove catch-labels (which actually should have been called "caught" labels but anyway)
2. allow `try`s without `catch`es
3. remove auxiliary `type` of labels and allow `delegate` to reference any label

WDYT?

Reply:
I just pushed the promised commits. The first one's commit message contains the question "Should the administrative instruction `caught` be removed from the right hand side of rethrow's execution step?", because I ran out of time and I won't be able to think about it until the end of next week.

Reply:
@aheejin: I just want to clarify that the commits I pushed are just so that it's clear what I meant with my comment above. These commits are independent from each other, so one could be accepted and the next one not without trouble, I think.

Reply:
Sorry for the delayed reply. The catch-less try was a typo, sorry. I didn't mean we have a catchless try. We can maybe consider it, but I personally have a preference against it, as I said in #131. What I tried to convey in that third example was that we can't make `delegate` target outer `catch`, because `delegte` can also target an outer `try-delegate`.